platform/linuxbsd/freedesktop_portal_desktop.cpp

```diff
@@ -126,6 +130,318 @@ uint32_t FreeDesktopPortalDesktop::get_appearance_color_scheme() {
+	dbus_message_unref(message);
+
+	if (!reply || dbus_error_is_set(&err)) {
+		ERR_PRINT(vformat("Failed to send DBus message: %s", err.message));
+		dbus_error_free(&err);
+		dbus_bus_remove_match(fd.connection, fd.path.utf8().get_data(), &err);
+		dbus_connection_unref(fd.connection);
+		return FAILED;
+	}
+
+	// Update signal path.
+	{
+		DBusMessageIter iter;
+		if (dbus_message_iter_init(reply, &iter)) {
+			if (dbus_message_iter_get_arg_type(&iter) == DBUS_TYPE_OBJECT_PATH) {
+				const char *new_path = nullptr;
+				dbus_message_iter_get_basic(&iter, &new_path);
+				if (String::utf8(new_path) != path) {
+					dbus_bus_remove_match(fd.connection, fd.path.utf8().get_data(), &err);
+					if (dbus_error_is_set(&err)) {
+						ERR_PRINT(vformat("Failed to remove DBus match: %s", err.message));
+						dbus_error_free(&err);
+						dbus_connection_unref(fd.connection);
+						return FAILED;
+					}
+					fd.path = String::utf8(new_path);
+					dbus_bus_add_match(fd.connection, fd.path.utf8().get_data(), &err);
+					if (dbus_error_is_set(&err)) {
+						ERR_PRINT(vformat("Failed to add DBus match: %s", err.message));
+						dbus_error_free(&err);
+						dbus_connection_unref(fd.connection);
+						return FAILED;
+					}
+				}
+			}
+		}
+	}
+	dbus_message_unref(reply);
+
+	MutexLock lock(file_dialog_mutex);
+	file_dialogs.push_back(fd);
+
+	return OK;
+}
+
+void FreeDesktopPortalDesktop::_thread_file_dialog_monitor(void *p_ud) {
+	FreeDesktopPortalDesktop *portal = (FreeDesktopPortalDesktop *)p_ud;
+
+	while (!portal->file_dialog_thread_abort.is_set()) {
+		{
+			MutexLock lock(portal->file_dialog_mutex);
+			for (int i = portal->file_dialogs.size() - 1; i >= 0; i--) {
+				bool remove = false;
+				{
+					FreeDesktopPortalDesktop::FileDialogData &fd = portal->file_dialogs.write[i];
+					if (fd.connection) {
+						while (true) {
+							DBusMessage *msg = dbus_connection_pop_message(fd.connection);
+							if (!msg) {
+								break;
+							} else if (dbus_message_is_signal(msg, "org.freedesktop.portal.Request", "Response")) {
+								DBusMessageIter iter;
+								if (dbus_message_iter_init(msg, &iter)) {
+									bool cancel = false;
+									Vector<String> uris;
+									file_chooser_parse_response(&iter, cancel, uris);
+
+									if (fd.callback.is_valid()) {
+										Variant v_status = !cancel;
+										Variant v_files = uris;
+										Variant *v_args[2] = { &v_status, &v_files };
+										fd.callback.call_deferredp((const Variant **)&v_args, 2);
+									}
+								}
+								dbus_message_unref(msg);
+
+								DBusError err;
+								dbus_error_init(&err);
+								dbus_bus_remove_match(fd.connection, fd.path.utf8().get_data(), &err);
+								dbus_error_free(&err);
+								dbus_connection_unref(fd.connection);
+								fd.connection = nullptr;
+								remove = true;
+								break;
+							}
+							dbus_message_unref(msg);
+						}
+						dbus_connection_read_write(fd.connection, 0);
```

Comment:
Clearing `fd.connection` here might feed `dbus_connection_read_write` on line 433 with a null, which results in a crash.